go-controller/pkg/ovn/ovn.go

```diff
@@ -633,95 +630,92 @@ func (oc *Controller) WatchPods() {
 // WatchNetworkPolicy starts the watching of network policy resource and calls
 // back the appropriate handler logic
 func (oc *Controller) WatchNetworkPolicy() {
-	go func() {
-		// track the retryNetworkPolicies map and every 30 seconds check if any pods need to be retried
-		for {
-			select {
-			case <-time.After(30 * time.Second):
-				oc.iterateRetryNetworkPolicies(false)
-			case <-oc.retryPolicyChan:
-				oc.iterateRetryNetworkPolicies(true)
-			case <-oc.stopChan:
-				return
-			}
-		}
-	}()
-
+	oc.triggerRetryObjs(oc.retryNetPolices.retryChan, oc.iterateRetryNetworkPolicies)
 	start := time.Now()
 	oc.watchFactory.AddPolicyHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			policy := obj.(*kapisnetworking.NetworkPolicy)
-			oc.initRetryPolicy(policy)
-			oc.checkAndSkipRetryPolicy(policy)
+			key := getPolicyNamespacedName(policy)
+			oc.retryNetPolices.initRetryObjWithAdd(policy, key)
+			oc.retryNetPolices.checkAndSkipRetryObj(key)
 			// If there is a delete entry this is a network policy being added
 			// with the same name as a previous network policy that failed deletion.
 			// Destroy it first before we add the new policy.
-			if retryEntry := oc.getPolicyRetryEntry(policy); retryEntry != nil && retryEntry.oldPolicy != nil {
+			if retryEntry := oc.retryNetPolices.getObjRetryEntry(key); retryEntry != nil && retryEntry.oldObj != nil {
 				klog.Infof("Detected stale policy during new policy add with the same name: %s/%s",
 					policy.Namespace, policy.Name)
-				if err := oc.deleteNetworkPolicy(retryEntry.oldPolicy, nil); err != nil {
-					oc.unSkipRetryPolicy(policy)
-					klog.Errorf("Failed to delete stale network policy %s, during add: %v",
-						getPolicyNamespacedName(policy), err)
-					return
+				knp, ok := retryEntry.oldObj.(*kapisnetworking.NetworkPolicy)
+				if ok {
```

Comment:
this should never fail and if somehow it does here, we are just ignoring it. I think you should just change this to a direct assertion and if it fails we would panic. There's something very wrong with our code if we are storing wrong types. So just do knp := retryEntry.oldObj.(*kapisnetworking.NetworkPolicy)